### PR TITLE
bugfix: notify p2p downloader to pull next piece after reporting

### LINKED ISSUE #1307: bug: dfget sleep much time to wait for pulling next downloading piece
### Ⅰ. Issue Description

dfget/supernode version: 1.0.0

dfget sleep much time to wait for pulling next downloading task.
It happens when there is only `cdn` has the available data.

### Ⅱ. Describe what happened

dfclient.log
```
2020-04-27 19:29:09.123 INFO sign:3947-1587986939.872 : pull piece task({"taskID":"5b1c283943dc1c8da0865907038b0401ce7f5f50b4a7bb2b8f8b7bcf7dda5d00","superNode":"127.0.0.1:8002","dstCid":"cdnnode:127.0.0.1~5b1c283943dc1c8da0865907038b0401ce7f5f50b4a7bb2b8f8b7bcf7dda5d00","range":"346030080-352321535","result":503,"status":701,"pieceSize":6291456,"pieceNum":55}) result:{"code":602,"msg":"{\"Code\":9,\"Msg\":\"peer should wait\"}"} and sleep 1.416s
2020-04-27 19:29:10.563 INFO sign:3947-1587986939.872 : downloading piece:{"taskID":"5b1c283943dc1c8da0865907038b0401ce7f5f50b4a7bb2b8f8b7bcf7dda5d00","superNode":"127.0.0.1:8002","dstCid":"cdnnode:127.0.0.1~5b1c283943dc1c8da0865907038b0401ce7f5f50b4a7bb2b8f8b7bcf7dda5d00","range":"377487360-383778815","result":503,"status":701,"pieceSize":6291456,"pieceNum":60}
2020-04-27 19:29:10.565 INFO sign:3947-1587986939.872 : pull piece task({"taskID":"5b1c283943dc1c8da0865907038b0401ce7f5f50b4a7bb2b8f8b7bcf7dda5d00","superNode":"127.0.0.1:8002","dstCid":"cdnnode:127.0.0.1~5b1c283943dc1c8da0865907038b0401ce7f5f50b4a7bb2b8f8b7bcf7dda5d00","range":"377487360-383778815","result":503,"status":701,"pieceSize":6291456,"pieceNum":60}) result:{"code":602,"msg":"{\"Code\":9,\"Msg\":\"peer should wait\"}"} and sleep 1.211s
```
### Ⅲ. Describe what you expected to happen

dfget should not sleep after a piece is downloaded.

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. start `supernode` with version 1.0.0
2. use `dfget` to download a large file into `supernode`, file size > 500MB
3. kill the `dfget server` process to make sure there is no other peer has the file
    ```bash
    pkill dfget
    ```
4. use `dfget` to download the same file again with a high speed, such as 400MB/s
    ```
    dfget -u "$url" -o /tmp/a.test --console --notbs -s 400MB
    ```
5. the problem happens

### Ⅵ. Environment:

- dragonfly version: 1.0.0
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by lowzj**
Signed-off-by: lowzj <zj3142063@gmail.com>

### Ⅱ. Does this pull request fix one issue?

fixes #1307 

should be merged into branch: `master`, `1.0.x`

### CODE REVIEW

File: dfget/core/downloader/p2p_downloader/p2p_downloader.go [hunk 1]
```diff
@@ -280,15 +286,9 @@ func (p2p *P2PDownloader) pullPieceTask(item *Piece) (
...
-		logrus.Infof("pull piece task(%+v) result:%s and sleep %.3fs", item, res, sleepTime.Seconds())
-		time.Sleep(sleepTime)
-
-		// gradually increase the sleep time, up to [800-1600]
-		if p2p.minTimeout < 800 {
-			p2p.minTimeout *= 2
-			p2p.maxTimeout *= 2
-		}
+		actual, expected := p2p.sleepInterval()
+		logrus.Infof("pull piece task(%+v) result:%s and sleep actual:%.3fs expected:%.3fs",
```
Comment (Starnop) [inline dfget/core/downloader/p2p_downloader/p2p_downloader.go:290]: Is it necessary to record expected sleep interval which is actually a random value.

File: dfget/core/downloader/p2p_downloader/client_stream_writer.go [hunk 2]
```diff
@@ -68,11 +71,12 @@ type ClientStreamWriter struct {
 }
 
 // NewClientStreamWriter creates and initialize a ClientStreamWriter instance.
-func NewClientStreamWriter(clientQueue queue.Queue, api api.SupernodeAPI, cfg *config.Config) *ClientStreamWriter {
+func NewClientStreamWriter(clientQueue, notifyQueue queue.Queue, api api.SupernodeAPI, cfg *config.Config) *ClientStreamWriter {
```
Comment (Starnop) [inline dfget/core/downloader/p2p_downloader/client_stream_writer.go:74]: I think we should abstract an interface for client writer, WDYT?

File: dfget/core/downloader/p2p_downloader/p2p_downloader.go (hunk 1 above)
Comment (lowzj) [reply dfget/core/downloader/p2p_downloader/p2p_downloader.go:290]: Yeah, it's more like a `debug` log, but I think it's necessary to output it in `info` level if `actual > expected`.

File: dfget/core/downloader/p2p_downloader/client_stream_writer.go (hunk 2 above)
Comment (lowzj) [reply dfget/core/downloader/p2p_downloader/client_stream_writer.go:74]: agreed, there is an ambitious plan to do this, but not in this pr.

Comment (Starnop) [reply dfget/core/downloader/p2p_downloader/client_stream_writer.go:74]: OK

**Comment by Starnop** [comment]
LGTM.

**Comment by Starnop** [comment]
OK, I think I understand what you want to do, and LGTM for this PR. 
However, it is difficult to understand simply with code only, I think some comments should be added for the `sleepInterval function` and `notifyQueue field`. 😄

**Comment by lowzj** [comment]
> Well, I think I understand what you want to do, and LGTM for this PR.
> However, it is difficult to understand simply with code only, I think some comments should be added for the `sleepInterval function` and `notifyQueue field`. 😄

DONE

**Pull request merged**